File: go/skynet.go
package main

import (
	"fmt"
	"sync"
	"time"
)

func skynet(c chan int, num int, size int, div int) {
	if size == 1 {
		c <- num
		return
	}

	rc := make(chan int)
	var sum int
	for i := 0; i < div; i++ {
		subNum := num + i*(size/div)
		go skynet(rc, subNum, size/div, div)
	}
	for i := 0; i < div; i++ {
		sum += <-rc
	}
	c <- sum
}

func skynetWait(c chan int64, num int, size int, div int, wg *sync.WaitGroup) {
	if size == 1 {
		c <- int64(num)
		wg.Done()
		return
	}

	rc := make(chan int64, div)
	var inner sync.WaitGroup
	inner.Add(div)
	sum := int64(0)
	for i := 0; i < div; i++ {
		subNum := num + i*(size/div)
		go skynetWait(rc, subNum, size/div, div, &inner)
	}
	inner.Wait()
	close(rc)
	for i := range rc {
		sum += i
	}
	c <- sum
	wg.Done()
}

func skynetSync(c chan int64, num int, size int, div int) {
	if size == 1 {
		c <- int64(num)
		return
	}

	rc := make(chan int64, 10)
	var sum int64
	for i := 0; i < div; i++ {
		subNum := num + i*(size/div)
		skynetSync(rc, subNum, size/div, div)
	}
	for i := 0; i < div; i++ {
		sum += <-rc
	}
	c <- sum
}

func skynetSyncWithoutChannel(num int, size int, div int) int64 {
	if size == 1 {
		return int64(num)
	}

	sum := int64(0)
	for i := 0; i < div; i++ {
		subNum := num + i*(size/div)
		sum += skynetSyncWithoutChannel(subNum, size/div, div)
	}
	return sum
}

func main() {
	c := make(chan int)
	start := time.Now()
	go skynet(c, 0, 1000000, 10)
	result := <-c
	took := time.Since(start)
	fmt.Printf("Result: %d in %d ms.\n", result, took.Nanoseconds()/1e6)

	buffer := make(chan int64, 10)
	start = time.Now()
	skynetSync(buffer, 0, 1000000, 10)
	result2 := <-buffer
	took = time.Since(start)
	fmt.Printf("Result sync: %d in %d ms.\n", result2, took.Nanoseconds()/1e6)

	start = time.Now()
	var wg sync.WaitGroup
	wg.Add(1)
	go skynetWait(buffer, 0, 1000000, 10, &wg)
	wg.Wait()
	result2 = <-buffer
	took = time.Since(start)
	fmt.Printf("Result using wait: %d in %d ms.\n", result2, took.Nanoseconds()/1e6)

	start = time.Now()
	result2 = skynetSyncWithoutChannel(0, 1000000, 10)
	took = time.Since(start)
	fmt.Printf("Result sync without channel: %d in %d ms.\n", result2, took.Nanoseconds()/1e6)

}

File: c_libmill/skynet.c
#include <libmill.h>
#include <stdio.h>

/*
 * C & libmill port of the goland skynet code.
 * 
 * Libmill (http://libmill.org/) is a library that brings golang like 
 * concurrency to C.
 * 
 * I found that while libmill's implementation is very simple the 
 * current approach to scheduling coroutines and the default stack size 
 * mean that spawning all 1111111 coroutines doesn't work, therefor
 * there are some differences between this and the original golang code.
 * 
 * Putting aside the question of whether such a benchmark represents a 
 * useful scenario I think it would be interesting to explore if the 
 * current limitations can be removed, baring in mind the limitations of 
 * the C runtime.
 * 
 * This is the first time I've done anything with libmill so the 
 * there may be faster ways to do things but I was interested in how 
 * comparible it was to golang.
 * 
 * You will find it you run the benchmark that it is slower than the 
 * golang version - mostly because of the time taken in mallocing so 
 * many large stacks.
 */

coroutine void skynet(chan c, int num, int size, int div) {
    if (size == 1) {
        chs(c, int64_t, num);
    } else {
        /* 
         * Slight change from go version which used an unbuffered 
         * channel.
         */
        chan rc = chmake(int64_t, div); 
        int64_t sum = 0;
        for(int64_t i = 0; i < div; i++) {
            int64_t sub_num = num + i * (size / div);
            /* 
             * Because of the way libmill schedules coroutines without
             * the following yield() more of the coroutines which start
             * additional coroutines will run before the ones which 
             * do not (size == 1). This would result in a lot of active 
             * coroutines each with a stack (default 256K although it 
             * can be changed to 16K without modifying libmill) and 
             * many mapped pages. Even with the yield you may find that 
             * you need to increase vm.max_map_count e.g.
             * 
             *     sudo sysctl -w vm.max_map_count=2000000
             */
            yield(); 
            go(skynet(rc, sub_num, size / div, div));
        }
        for(int64_t i = 0; i < div; i++) {
            sum += chr(rc, int64_t);
        }
        chclose(rc);
        chs(c, int64_t, sum);
    }
}

int main(int argc, char **argv) {
    chan c = chmake(int64_t, 0);
    int64_t start_ms = now();
    go(skynet(c, 0, 1000000, 10));
    int64_t result = chr(c, int64_t);
    int64_t end_ms = now();
    chclose(c);
    printf("Result: %ld in %d ms.\n", result, end_ms - start_ms);
    return 0;
}
